tests/py3/test_lookup.py

```diff
@@ -365,3 +365,37 @@ def test_required_args(target, args, kwargs):
     # Mostly checking that `self` (and only self) is correctly excluded
     st.builds(target, *map(st.just, args),
               **{k: st.just(v) for k, v in kwargs.items()}).example()
+
+
+@pytest.mark.skipif(sys.version_info[:2] < (3, 6),
+                    reason='arguments not resolved by old typing module')
+def test_cannot_resolve_bare_forward_reference():
+    # loop instead of parametrize because typing.Type was added in 3.6
+    for thing in [typing.Optional, typing.List, typing.Type]:
+        with pytest.raises(InvalidArgument):
+            st.from_type(thing['SomeReference']).example()
+
+
+class Tree:
+    def __init__(self,
+                 left: typing.Optional['Tree'],
+                 right: typing.Optional['Tree']):
+        self.left = left
+        self.right = right
+
+    def __repr__(self):
+        return 'Tree({}, {})'.format(self.left, self.right)
+
+    @property
+    def size(self):
+        return sum([
+            1,
+            self.left.size if self.left else 0,
+            self.right.size if self.right else 0
+        ])
+
+
+@pytest.mark.skipif(sys.version_info[:2] < (3, 6),
+                    reason='arguments not resolved by old typing module')
```

Comment:
If we're only going to support this on >= 3.6 (which I'm 100% on board with) it would be nice if the error message we raise in this case were more specific about supported versions.